activejob/lib/active_job/async_job.rb

```diff
@@ -0,0 +1,98 @@
+require 'concurrent'
+require 'thread_safe'
+
+module ActiveJob
+  # == Active Job Async Job
+  #
+  # When enqueueing jobs with Async Job each job will be executed
+  # asynchronously on a +concurrent-ruby+ thread pool. All job data
+  # is retained in memory. Because job data is not saved to a persistent
+  # datastore there is no additional infrastructure needed and most jobs
+  # will process very quickly. The lack of persistence, however, means
+  # that all unprocessed jobs will be lost on application restart. This
+  # makes in-memory queue adapters unsuitable for most production
+  # environments but excellent for development and testing.
+  #
+  # Read more about Concurrent Ruby {here}[https://github.com/ruby-concurrency/concurrent-ruby].
+  #
+  # To use Async Job set the queue_adapter config to +:async+.
+  #
+  #   Rails.application.config.active_job.queue_adapter = :async
+  #
+  # Async Job supports job queues specified with +queue_as+. Queues
+  # are implemented as individual thread pools. Queues will be created
+  # automatically as needed using a default configuration. Should there be a
+  # need to customize the thread pool behind one or more queues, any
+  # concurrent-ruby thread pool can be injected. Simply create the thread
+  # pool directly then call the +create_queue+ method passing the thread
+  # pool as the second parameter:
+  #
+  #   thread_pool = Concurrent::FixedThreadPool.new(10, max_queue: 100, fallback_polcy: :caller_runs)
+  #   ActiveJob::AsyncJob.create_queue(:my_queue, thread_pool)
+  #
+  # The +create_queue+ method must be called during application initialization,
+  # before any jobs are post to the queue. Attempting to create a queue that
+  # already exists will raise an error.
+  class AsyncJob
+
+    DEFAULT_EXECUTOR_OPTIONS = {
+      min_threads:     [2, Concurrent.processor_count].max,
+      max_threads:     Concurrent.processor_count * 10,
+      auto_terminate:  true,
+      idletime:        60, # 1 minute
+      max_queue:       0, # unlimited
+      fallback_policy: :caller_runs # shouldn't matter -- 0 max queue
+    }.freeze
+
+    QUEUES = ThreadSafe::Cache.new do |hash, queue_name|
+      hash[queue_name] = ActiveJob::AsyncJob.create_thread_pool
+    end
+    private_constant :QUEUES
+
+    # Raised when an attempt is made to create a queue that already exists.
+    class QueueCreationError < ArgumentError; end
+
+    class << self
+      # Force all jobs to run synchronously when testing the activejob gem.
```

Comment:
@kaspth I apologize if I'm missing something, but it appears that the file `test/integration/queuing_test.rb` currently tests the asynchronous behavior of all adapters. It posts jobs with `perform_later` then uses timers (such as `wait_for_jobs_to_finish_for(2.seconds)`) to verify that the jobs post. It looks like the adapter tests put all the adapters in synchronous mode, the existing integration tests put the adapters in asynchronous mode, and the Rakefile controls the environment appropriately for each set of tests. However, I didn't notice this earlier so I have not setup the necessary test helper for making the integration tests work with the Async Job. The integration tests (which do not run with `rake test`) are failing. I'll fix that.